### PR TITLE
chore: Add Textfields components

### LINKED ISSUE #159: Add TextField

### PR DESCRIPTION
**Pull request by soulcramer**
## 🧑‍ Changes description

Add Textfields components

## 📸 Screenshots

Integrated in the review files

## 👋 Other info
close #159

### CODE REVIEW

File: spark/src/main/kotlin/com/adevinta/spark/components/textfields/DefaultSparkTextFieldColors.kt [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.Immutable
+import androidx.compose.runtime.State
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.rememberUpdatedState
+import androidx.compose.ui.graphics.Color
+import com.adevinta.spark.SparkTheme
+
+@Composable
+internal fun sparkOutlinedTextFieldColors(
```
Comment (SimonMarquis) [inline spark/src/main/kotlin/com/adevinta/spark/components/textfields/DefaultSparkTextFieldColors.kt:43]: Shouldn't they be part of the public API for consumer to use?

File: spark/src/main/kotlin/com/adevinta/spark/components/textfields/DefaultSparkTextFieldColors.kt [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+        return rememberUpdatedState(if (isError) errorCursorColor else cursorColor)
+    }
+
+    /**
+     * Represents the colors used for text selection in this text field.
+     */
+    internal val selectionColors: TextSelectionColors
+        @Composable get() = textSelectionColors
+
+    override fun equals(other: Any?): Boolean {
```
Comment (SimonMarquis) [inline spark/src/main/kotlin/com/adevinta/spark/components/textfields/DefaultSparkTextFieldColors.kt:305]: Any particular reason to not take advantage of Kotlin data classes here?

File: spark/src/main/kotlin/com/adevinta/spark/components/textfields/MultilineTextField.kt [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+    value: String,
+    onCancelClick: () -> Unit,
+): (@Composable () -> Unit)? = when {
+    isError -> {
+        {
+            Icon(
+                modifier = Modifier
+                    .rotate(180f),
+                imageVector = Icons.Default.Info,
+                contentDescription = null, // We should walready knnow that we're on error
```
Comment (SimonMarquis) [inline spark/src/main/kotlin/com/adevinta/spark/components/textfields/MultilineTextField.kt:161]:
```suggestion
                contentDescription = null, // We should already know that we're on error
```

File: spark/src/main/kotlin/com/adevinta/spark/components/textfields/SelectTextField.kt [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+import androidx.compose.ui.tooling.preview.PreviewParameter
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.window.PopupProperties
+import com.adevinta.spark.ExperimentalSparkApi
+import com.adevinta.spark.PreviewTheme
+import com.adevinta.spark.tools.preview.ThemeProvider
+import com.adevinta.spark.tools.preview.ThemeVariant
+
+/**
+ * Outlined text input to get an input value from a list of elements selectable thought a dropdown by the user.
```
Comment (SimonMarquis) [inline spark/src/main/kotlin/com/adevinta/spark/components/textfields/SelectTextField.kt:62]:
```suggestion
 * Outlined text input to get an input value from a list of elements selectable through a dropdown by the user.
```

File: spark/src/main/kotlin/com/adevinta/spark/components/textfields/SelectTextField.kt [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+            onDismissRequest = onDismissRequest,
+            modifier = Modifier.exposedDropdownSize(),
+            properties = properties,
+            content = dropdownContent,
+        )
+    }
+}
+
+/**
+ * Outlined text input to get an input value from a list of elements selectable thought a dropdown by the user.
```
Comment (SimonMarquis) [inline spark/src/main/kotlin/com/adevinta/spark/components/textfields/SelectTextField.kt:165]:
```suggestion
 * Outlined text input to get an input value from a list of elements selectable through a dropdown by the user.
```

File: spark/src/main/kotlin/com/adevinta/spark/components/textfields/SparkTextField.kt [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+
+@Stable
+data class TextFieldCharacterCounter(
+    val count: Int,
+    val maxCharacter: Int,
+)
+
+@Composable
+private fun Label(text: String?) {
+    if (text?.take(30) != null) {
```
Comment (SimonMarquis) [inline spark/src/main/kotlin/com/adevinta/spark/components/textfields/SparkTextField.kt:322]: This creates unnecessary strings. This in fact only checks if `text` is not null.

File: spark/src/main/kotlin/com/adevinta/spark/components/textfields/SparkTextField.kt [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+        Box {
+            Text(text = text)
+        }
+    }
+}
+
+@Composable
+private fun PlaceHolder(text: String?) {
+    if (text?.take(40) != null) {
```
Comment (SimonMarquis) [inline spark/src/main/kotlin/com/adevinta/spark/components/textfields/SparkTextField.kt:331]: same

File: spark/src/main/kotlin/com/adevinta/spark/components/textfields/SparkTextFieldImpl.kt [hunk 8]
```diff
@@ -0,0 +1,537 @@
...
+                        content = counter,
+                    )
+                }
+            }
+        }
+
+        // Developers need to handle invalid input manually. But since we don't provide error
+        // message slot API, we can set the default error message in case developers forget about
+        // it.
+//        val defaultErrorMessage = error ?: "une erreur est survenue"
```
Comment (SimonMarquis) [inline spark/src/main/kotlin/com/adevinta/spark/components/textfields/SparkTextFieldImpl.kt:179]: To be removed?

File: spark/src/main/kotlin/com/adevinta/spark/components/textfields/SparkTextFieldImpl.kt [hunk 9]
```diff
@@ -0,0 +1,537 @@
...
+    }
+    if (typography != null) ProvideTextStyle(typography, colorAndEmphasis) else colorAndEmphasis()
+}
+
+/**
+ * Layout of the leading and trailing icons and the text field, label and placeholder in
+ * [SparkTextFieldLayout].
+ * It doesn't use Row to position the icons and middle part because label should not be
+ * positioned in the middle part.
+\ */
```
Comment (SimonMarquis) [inline spark/src/main/kotlin/com/adevinta/spark/components/textfields/SparkTextFieldImpl.kt:269]:
```suggestion
 */
```

File: spark/src/main/kotlin/com/adevinta/spark/components/textfields/DefaultSparkTextFieldColors.kt (hunk 1 above)
Comment (soulcramer) [reply spark/src/main/kotlin/com/adevinta/spark/components/textfields/DefaultSparkTextFieldColors.kt:43]: Nope as we don't want them to customize the textfield colors

File: spark/src/main/kotlin/com/adevinta/spark/components/textfields/DefaultSparkTextFieldColors.kt (hunk 2 above)
Comment (soulcramer) [reply spark/src/main/kotlin/com/adevinta/spark/components/textfields/DefaultSparkTextFieldColors.kt:305]: I can't answer you for this as it's a copy of the OutlineTextfield from Material to support helper text and char counter :-/

**Review by SimonMarquis — APPROVED** [review]

**Review by SimonMarquis — APPROVED** [review]

**Pull request merged**